### PR TITLE
"crow::multipart::message throws when the boundary is empty or malformed

### PR DESCRIPTION
**Pull request by Iuliean**
This related to issue #916 where the multipart message causes a segmentation fault when a multipart request with an empty boundary is received. 

I added two throws to indicate whenever the boundary is empty or it could not be found/differs from the boundary inside the body.

I hope these changes are enough to prevent the code from being unable to parse the headers thus leaving to the segmentation faults.

I also added two new tests that aim to check that these exceptions are thrown whenever the boundary is either missing or not matching

### CODE REVIEW

**Comment by The-EDev** [comment]
Quick question about this, I understand that raising an exception is a valid solution, but if this exception goes unhandled then we end up with a `500` response to the client, rather than the appropriate `400` response. 

Would it be appropriate to just respond with a `400` rather than raise an exception and have the handler deal with it? What are your thoughts?

**Comment by Iuliean** [comment]
I have thought about automatically responding with an error code when this is detected but i mostly didn't know where to look to see how that can be achieved.

If you point me in the right direction i'll try to make the necessary changes. 

One counter argument i can give to what you said is. If the user does not handle the error in his handler returning 500 seems completely fine to me since an internal error did happen but the user did not check for it so it is unknown (to me that seems completely reasonable). Any other unhandled error would have the same result i assume.

But if you wish to automatically respond with this. The easiest way would probably be to have a BadRequestException that can be caught in the method that handles each handler. But let me know if there is a better way

**Comment by The-EDev** [comment]
> One counter argument i can give to what you said is. If the user does not handle the error in his handler returning 500 seems completely fine to me since an internal error did happen but the user did not check for it so it is unknown (to me that seems completely reasonable). Any other unhandled error would have the same result i assume.

since the exception is the result of an error within the request itself, and since the only fix is to actually change the request, then it should be a `400`, `500` should be used if the error is the result of a flaw in the server itself, which doesn't match this case.

> The easiest way would probably be to have a BadRequestException that can be caught in the method that handles each handler. But let me know if there is a better way

I agree, I think this would be ideal, Crow handles exceptions [here](https://github.com/CrowCpp/Crow/blob/bd61a7d1715f090b0b1b0c35e9e7f697aa42591b/include/crow/routing.h#L1807). I think modifying the default exception handler to recognize a bad request exception is a valid solution.

**Comment by Iuliean** [comment]
Yes i agree, i'll make the required changes as soon as i can

**Comment by Iuliean** [comment]
One question i have is. 
Looking at the code for the `request` class, it seems that it just a holder of all the information needed to then build a multipart message.

What if request itself could handle the multipart stuff, then when the request itself is being built the error would be caught  way before it enters the user's handler. Wouldn't that be preferred ?
I mean you lose the "lazy loading" part of it where if you don't parse you have 0 performace loss but all errors are caught before they become an issue.
Also ,maybe this would be a very rare case, it could happen that you respond to invalid requests sometimes. Say you check for some url values first and based on that you deduce there is no need to parse it, thus an otherwise invalid request is treated as normal.

**Comment by The-EDev** [comment]
> What if request itself could handle the multipart stuff, then when the request itself is being built the error would be caught way before it enters the user's handler. Wouldn't that be preferred ?

well it's a design decision, as far as I can tell, the design is that `request` and `response` are only meant to hold data, with any logic being separated from them.

> Say you check for some url values first and based on that you deduce there is no need to parse it, thus an otherwise invalid request is treated as normal.

wouldn't putting this kind of logic in the handler be a good solution?

File: include/crow/http_request.h [hunk 1]
```diff
@@ -31,6 +32,15 @@ namespace crow // NOTE: Already documented in "crow/app.h"
...
 
+    struct bad_request : public std::runtime_error
+    {
+        bad_request(const std::string& what_arg)
+            : std::runtime_error(what_arg) {}
+
+        bad_request(const char* what_arg)
+            : std::runtime_error(what_arg) {}
+    };
+
```
Comment (The-EDev) [inline include/crow/http_request.h:43]: apologies, I just noticed this, wouldn't it be best to have this be either in `utility.h` or another header file specifically for exceptions?

**Review by The-EDev — CHANGES_REQUESTED** [review]

File: include/crow/http_request.h (hunk 1 above)
Comment (Iuliean) [reply include/crow/http_request.h:43]: well i did not see any exception specific header. i am fine with either utility.h or an "exceptions.h` or whatever name you wish :)

Comment (The-EDev) [reply include/crow/http_request.h:43]: `exceptions.h` sounds good!

File: include/crow/http_request.h [hunk 2]
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <stdexcept>
```
Comment (The-EDev) [inline include/crow/http_request.h:3]: I'm guessing this is no longer necessary.

File: include/crow/routing.h [hunk 3]
```diff
@@ -1813,6 +1813,11 @@ namespace crow // NOTE: Already documented in "crow/app.h"
             {
                 throw;
             }
+            catch (const bad_request& e)
+            {
+                res = response (400);
+                res.body = e.what();
+            }
```
Comment (The-EDev) [inline include/crow/routing.h:1820]: I think the include for `exceptions.h` is missing and it's causing the tests to fail.

**Review by The-EDev — CHANGES_REQUESTED** [review]

**Review by The-EDev — APPROVED** [review]
Approved, Thank you for your help!
